python/feathub/processors/spark/dataframe_builder/join_utils.py

```diff
@@ -0,0 +1,145 @@
+from feathub.table.table_descriptor import TableDescriptor
+from pyspark.sql import DataFrame as NativeSparkDataFrame, SparkSession
+from pyspark.sql.types import DataType
+
+
+class JoinFieldDescriptor:
+    """
+    Descriptor of the join field.
+    """
+
+    def __init__(
+        self,
+        field_name: str,
+        field_data_type: Optional[DataType] = None,
+    ):
+        """
+        :param field_name: The name of the field to join.
+        :param field_data_type: Optional. If it is not None, the field is cast to the
+                                given type. Otherwise, use its original type.
+        """
+        self.field_name = field_name
+        self.field_data_type = field_data_type
+
+    @staticmethod
+    def from_table_descriptor_and_field_name(
+        table_descriptor: TableDescriptor, field_name: str
+    ) -> "JoinFieldDescriptor":
+        feature = table_descriptor.get_feature(field_name)
+        transform = feature.transform
+
+        if (
+            not isinstance(table_descriptor, SlidingFeatureView)
+            or not isinstance(transform, SlidingWindowTransform)
+            or table_descriptor.config.get(ENABLE_EMPTY_WINDOW_OUTPUT_CONFIG)
+        ):
+            return JoinFieldDescriptor(field_name, to_spark_type(feature.dtype))
+        else:
+            raise FeathubException(
+                "Spark processor does not support sliding window currently."
+            )
+
+    @staticmethod
+    def from_field_name(field_name: str) -> "JoinFieldDescriptor":
+        return JoinFieldDescriptor(field_name)
+
+    def __eq__(self, other: object) -> bool:
+        return (
+            isinstance(other, self.__class__)
+            and self.field_name == other.field_name
+            and self.field_data_type == other.field_data_type
+        )
+
+    def __hash__(self) -> int:
+        return hash(
+            (
+                self.field_name,
+                self.field_data_type,
+            )
+        )
+
+
+def temporal_join(
+    spark: SparkSession,
+    left: NativeSparkDataFrame,
+    right: NativeSparkDataFrame,
+    keys: Sequence[str],
+) -> NativeSparkDataFrame:
+    """
+    Temporal join the right dataframe to the left dataframe.
+
+    :param spark: The SparkSession.
+    :param left: The left dataframe.
+    :param right: The right dataframe.
+    :param keys: The join keys.
+    :return: The joined dataframe.
+    """
+
+    overlapping_field_names = list(
+        set(right.schema.fieldNames()).intersection(set(left.schema.fieldNames()))
+    )
+
+    right_aliased = _get_field_aliased_right_dataframe(right, overlapping_field_names)
+
+    left.createOrReplaceTempView("left_table")
+    right_aliased.createOrReplaceTempView("right_table")
+
+    escaped_keys = [f"`{k}`" for k in keys]
+    join_predicates = " and ".join(
+        [f"left_table.`{k}` = right_table.`right.{k}`" for k in keys]
+    )
+    result_table = spark.sql(
+        f"""
```

Comment:
Does Spark provide table API to do this join? If it does, we want to avoid SQL as much as possible. The only reason that Flink processors use SQL to do temporal join is that Flink only support temporal join with SQL API.